multi_token_farm/src/lib.rs

```diff
@@ -0,0 +1,785 @@
+        require!(
+            self.pair_address_for_accepted_lp_token()
+                .contains_key(&token),
+            "No Pair Address for given LP token"
+        );
+        require!(
+            self.pair_address_for_accepted_lp_token()
+                .get(&token)
+                .unwrap()
+                == address,
+            "Address does not match Lp token equivalent"
+        );
+        self.pair_address_for_accepted_lp_token().remove(&token);
+        Ok(())
+    }
+
+    #[payable("*")]
+    #[endpoint]
+    fn enterFarm(
+        &self,
+        #[payment_token] token_in: TokenIdentifier,
+        #[payment] amount: Self::BigUint,
+    ) -> SCResult<SftTokenAmountPair<Self::BigUint>> {
+        require!(self.is_active(), "Not active");
+        require!(!self.farm_token_id().is_empty(), "No issued farm token");
+        let farm_contribution = self.get_farm_contribution(&token_in, &amount)?;
+        require!(farm_contribution > 0, "Cannot farm with amount of 0");
+
+        let is_first_provider = self.is_first_provider();
+        let farming_pool_token_id = self.farming_pool_token_id().get();
+        let mut liquidity = self.add_liquidity(
+            farm_contribution.clone(),
+            farming_pool_token_id,
+            token_in.clone(),
+        )?;
+        let farm_attributes = FarmTokenAttributes::<Self::BigUint> {
+            farmed_token_id: token_in,
+            total_farmed_tokens: amount,
+            total_initial_worth: farm_contribution,
+            total_amount_liquidity: liquidity.clone(),
+            entering_epoch: self.blockchain().get_block_epoch(),
+        };
+
+        // Do the actual permanent lock of first minimum liquidity
+        // only after the token attributes are crafted for the user.
+        if is_first_provider {
+            liquidity -= Self::BigUint::from(self.minimum_liquidity_farm_amount());
+        }
+
+        // This 1 is necessary to get_esdt_token_data needed for calculateRewardsForGivenPosition
+        let farm_tokens_to_create = &liquidity + &Self::BigUint::from(1u64);
+        let farm_token_id = self.farm_token_id().get();
+        self.create_farm_tokens(&farm_token_id, &farm_tokens_to_create, &farm_attributes);
+        let farm_token_nonce = self.farm_token_nonce().get();
+
+        self.send_tokens(
+            &farm_token_id,
+            farm_token_nonce,
+            &liquidity,
+            &self.blockchain().get_caller(),
+        );
+
+        Ok(SftTokenAmountPair {
+            token_id: farm_token_id,
+            token_nonce: farm_token_nonce,
+            amount: liquidity,
+        })
+    }
+
+    #[payable("*")]
+    #[endpoint]
+    fn exitFarm(
+        &self,
+        #[payment_token] payment_token_id: TokenIdentifier,
+        #[payment] liquidity: Self::BigUint,
+    ) -> SCResult<MultiResult2<TokenAmountPair<Self::BigUint>, TokenAmountPair<Self::BigUint>>>
+    {
+        //require!(self.is_active(), "Not active");
```

Comment:
please do not leave commented lines. If you have commented lines, add a TODO as well.